console/v1/types_console_sample.go

```diff
@@ -0,0 +1,193 @@
+// +kubebuilder:validation:Enum:=GitImport;ContainerImport
+type ConsoleSampleSourceType string
+
+const (
+	// A sample that let the user import code from a git repository.
+	GitImport ConsoleSampleSourceType = "GitImport"
+	// A sample that let the user import a container image.
+	ContainerImport ConsoleSampleSourceType = "ContainerImport"
+)
+
+// ConsoleSampleSource is the actual sample definition and can hold different sample types.
+// Unsupported sample types will be ignored in the web console.
+// +union
+// +kubebuilder:validation:XValidation:rule="self.type == 'GitImport' ? has(self.gitImport) : !has(self.gitImport)",message="source.gitImport is required when source.type is GitImport, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="self.type == 'ContainerImport' ? has(self.containerImport) : !has(self.containerImport)",message="source.containerImport is required when source.type is ContainerImport, and forbidden otherwise"
+type ConsoleSampleSource struct {
+	// type of the sample, currently supported: "GitImport";"ContainerImport"
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:="GitImport";"ContainerImport"
+	Type ConsoleSampleSourceType `json:"type"`
+
+	// gitimport let the user import code from a git repository.
+	// +unionMember
+	// +optional
+	GitImport *ConsoleSampleGitImportSource `json:"gitImport,omitempty"`
+
+	// containerimport let the user import a container image.
+	// +unionMember
+	// +optional
+	ContainerImport *ConsoleSampleContainerImportSource `json:"containerImport,omitempty"`
+}
+
+// ConsoleSampleGitImportSource let the user import code from a public git repository.
+type ConsoleSampleGitImportSource struct {
+	// url of the Git repository. Only public HTTPS repositories on GitHub, GitLab
+	// and Bitbucket are currently supported:
+	//
+	//   - https://github.com/<org>/<repository>
+	//   - https://gitlab.com/<org>/<repository>
+	//   - https://bitbucket.org/<org>/<repository>
+	//
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=100
+	// +kubebuilder:validation:Pattern=`^https:\/\/(github.com|gitlab.com|bitbucket.org)\/[a-zA-Z0-9-]+\/[a-zA-Z0-9-]+(.git)?$`
+	URL string `json:"url"`
+	// revision to clone a specific branch, tag or commit SHA, etc.
+	// +optional
+	Revision string `json:"revision"`
+	// contextDir, a subfolder in the repo to build the component.
+	// +optional
+	// +kubebuilder:validation:Pattern=`^/`
+	ContextDir string `json:"contextDir"`
+}
+
+// ConsoleSampleContainerImportSource let the user import a container image.
+type ConsoleSampleContainerImportSource struct {
+	// Reference to a container image like openshift/hello-openshift,
+	// docker.io/openshift/hello-openshift or quay.io/<username>/<image-name>
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=200
```

Comment:
Is 200 characters enough here? How long is a standard SHA for referring to an image? Are there any limits on the number of namespaces or length of the namespaces in the URL?

Reply:
Its enough for the common scenarios. Changed it to 256 since we have used this for the git URL as well which is normally shorter then a image ref with a SHA.